pkg/storage/internal/actions/funder/sql.go

```diff
@@ -32,5 +44,132 @@ func NewSQL(logger *slog.Logger, db *gorm.DB, feeModel defs.FeeModel) *SQL {
 // @param minimumDesiredUTXOValue - the minimum value of UTXO in basket #TakeFromBasket
 // @param userID - the user ID.
 func (f *SQL) Fund(ctx context.Context, targetSat int64, currentTxSize int64, numberOfDesiredUTXOs int, minimumDesiredUTXOValue uint64, userID int) (*actions.FundingResult, error) {
-	panic("not implemented")
+	txSize, err := to.UInt64(currentTxSize)
+	if err != nil {
+		return nil, fmt.Errorf("invalid currentTxSize: %w", err)
+	}
+
+	txSats, err := to.UInt64(targetSat)
+	if err != nil {
+		return nil, fmt.Errorf("invalid targetSat: %w", err)
+	}
+
+	utxos := f.loadUTXOs(ctx, userID)
+
+	result, err := f.allocate(txSats, txSize, utxos)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fund transaction: %w", err)
+	}
+	return result, nil
+
+}
+
+func (f *SQL) loadUTXOs(ctx context.Context, userID int) iter.Seq2[*models.UserUTXO, error] {
+	batches := seqerr.ProduceWithArg(
+		func(page *paging.Page) ([]*models.UserUTXO, *paging.Page, error) {
+			utxos, err := f.utxoRepository.FindAllUTXOs(ctx, userID, page)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to load utxos: %w", err)
+			}
+			page.Next()
+			return utxos, page, nil
+		},
+		&paging.Page{
+			Limit:  1000,
+			SortBy: "satoshis",
+		})
+
+	return seqerr.FlattenSlices(batches)
+}
+
+func (f *SQL) allocate(sats uint64, size uint64, utxos iter.Seq2[*models.UserUTXO, error]) (*actions.FundingResult, error) {
+	collector := newCollector(sats, size, f.feeCalculator)
+	for utxo, err := range utxos {
+		if err != nil {
+			return nil, fmt.Errorf("failed to allocate utxo: %w", err)
+		}
+
+		err = collector.Allocate(utxo)
+		if err != nil {
+			return nil, fmt.Errorf("failed to allocate utxo: %w", err)
+		}
+
+		if collector.IsFunded() {
+			break
+		}
+	}
+
+	return collector.GetResult()
```

Comment:
Consider this - maybe it would more straightforward. (And it doesn't need to call IsFunded twice)
```suggestion
		if collector.IsFunded() {
			return collector.result, nil
		}
	}

	return nil, errfunder.NotEnoughFunds
```

Reply:
I already have next steps, and later I need to calculate change and change count, I'm not sure yet if I want to do that during processing, that's why I prefer to keep this method.